shell-completion/hledger-completion.bash.m4

```diff
@@ -1,102 +1,102 @@
+# the rest of the session and completion for other programs.
 
-# Note: grep "^$wordToComplete" is (functional) not safe to use if the word
-# contains regex special chars. But it might be no problem because of
-# COMP_WORDBREAKS.
-
-# Note: compgen and compopt is pretty complicated. Piping to
-# grep "^$wordToComplete"
-# seems like a hack - I'd rather use
-# compgen ... -- "$wordToComplete"
-# But what options to use? I don't want to use -W because it may exceed the
-# maximum command line length. -C "cat file" is not working either. It would be
-# best if compgen could read from stdin but it does not.
-
-# Note: Working with bash arrays is nasty compared to editing a text file.
-# Consider for example grepping an array or mapping a substitution on it.
-# Therefore, we create temp files in RAM for completion suggestions (see below).
+_hledger_completion_function() {
+    # Current treatment for special characters:
+    # - exclude colon (:) from COMP_WORDBREAKS
+    # - use comptop -o filenames to escape the rest
+    COMP_WORDBREAKS=${COMP_WORDBREAKS//:}
+    compopt -o filenames
 
-readonly _HLEDGER_COMPLETION_TEMPDIR=$(mktemp -d)
+    local wordToComplete=$2
+    local subcommand
+    local subcommandOptions
+    local i
 
-_hledger_completion_function() {
-    #declare cmd=$1
-    declare wordToComplete=$2
-    declare precedingWord=$3
-
-    declare subcommand
-    for subcommand in "${COMP_WORDS[@]}"; do
-	if grep -Fxqe "$subcommand" "$_HLEDGER_COMPLETION_TEMPDIR/commands.txt"; then
-	    COMPREPLY+=( $(grep -h "^$wordToComplete" -- "$_HLEDGER_COMPLETION_TEMPDIR/options-$subcommand.txt") )
-	    break
-	fi
-	subcommand=
+    for (( i=1; i<${#COMP_WORDS[@]}; i++ )); do
+        subcommand=${COMP_WORDS[i]}
+        if ! grep -Fxqe "$subcommand" <<< "$_hledger_complist_commands"; then
+            subcommand=
+            continue
+        fi
+        # There could be other commands begining with $subcommand, e.g.:
+        # $subcommand == reg --> register, register-match,
+        # $subcommand == bal --> balance, balancesheet, balancesheetequity, etc.
+        # Do not ignore them!
+        if [[ $subcommand == "$wordToComplete" ]] && ((i == COMP_CWORD)); then
+            local subcommandMatches
+            subcommandMatches=$(grep -c "^$wordToComplete" <<< "$_hledger_complist_commands")
+            if ((subcommandMatches > 1)); then
+                subcommand=
+                break
+            else
+                _hledger_compreply "$subcommand"
+                return 0
+            fi
+        fi
+        # Replace dashes with underscores and use indirect expansion
+        subcommandOptions=_hledger_complist_options_${subcommand//-/_}
+        _hledger_compreply "$(_hledger_compgen "${!subcommandOptions}")"
+        break
     done
 
     if [[ -z $subcommand ]]; then
+        _hledger_compreply_optarg && return
 
-	declare completeFiles filenameSoFar
-	case $precedingWord in
-	    -f|--file|--rules-file)
```

Comment:
Sorry, if I overlooked it, but how are filename completions now handled? And is `-f` still working with completion?

Reply:
Hi, filenames are completed together with all other options *requiring* an argument by _hledger_compreply_optarg(). It handles completion for -f, --file, --pivot, etc. We call it twice, so we can handle option argument completion before and after the subcommand. And because we need to call it twice I moved it out in a separate function.